dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/security/apikey/ApiTokenServiceImpl.java

```diff
@@ -119,26 +119,34 @@ public void delete( ApiToken apiToken )
         // Invalidate cache here or let cache expire ?
     }
 
+    @Override
     public ApiToken initToken( ApiToken token )
     {
-        String randomSecureToken = getRandomSecureToken( 24 ).replaceAll( "[-_]", "x" );
-
-        CRC32 crc = new CRC32();
-        byte[] bytes = randomSecureToken.getBytes();
-        crc.update( bytes, 0, bytes.length );
-        long checksum = crc.getValue();
+        Preconditions.checkNotNull( token );
+        Preconditions.checkNotNull( token.getType() );
 
-        randomSecureToken = "d2p_" + randomSecureToken + checksum;
-
-        token.setKey( randomSecureToken );
+        ApiToken.ApiTokenType tokenType = ApiToken.ApiTokenType.values()[token.getType()];
```

Comment:
I think we should avoid given the `ordinal` a semantic and instead use an explicit external key of some sort.

Reply:
I agree ordinals are not preferred, I changed it to use Hibernate persisted enum types instead now.